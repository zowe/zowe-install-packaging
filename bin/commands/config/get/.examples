zwe config get -p path-of-configuration -c /path/to/zowe.yaml
zwe config get -p .zowe.certificate -c /path/to/zowe.yaml
zwe config get -p .zowe.components[] -c /path/to/zowe.yaml
zwe config get -p .zowe.components.gateway -c /path/to/zowe.yaml
zwe config get -p .zowe.components.gateway.enabled -c /path/to/zowe.yaml