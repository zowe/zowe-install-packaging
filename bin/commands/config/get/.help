Return value of a configuration defined in YAML configuration.
This command requires zowe.useConfigmgr=true or --configmgr to be used.
