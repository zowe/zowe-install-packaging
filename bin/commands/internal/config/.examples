    zwe internal config get -p path-of-configuration -c /path/to/zowe.yaml

    zwe internal config set -p path-of-configuration -e new-value-of-configuration -c /path/to/zowe.yaml
    