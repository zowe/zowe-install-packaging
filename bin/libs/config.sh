#!/bin/sh

################################################################################
# This program and the accompanying materials are made available under the terms of the
# Eclipse Public License v2.0 which accompanies this distribution, and is available at
# https://www.eclipse.org/legal/epl-v20.html
#
# SPDX-License-Identifier: EPL-2.0
#
# Copyright IBM Corporation 2021
################################################################################

################################################################################
# @internal 

###############################
# Convert instance.env to zowe.yaml file
convert_instance_env_to_yaml() {
  instance_env="${1}"
  zowe_yaml="${2}"

  # we need node for following commands
  ensure_node_is_on_path 1>/dev/null 2>&1

  if [ -z "${zowe_yaml}" ]; then
    node "${ROOT_DIR}/bin/utils/config-converter/src/cli.js" env yaml "${instance_env}"
  else
    node "${ROOT_DIR}/bin/utils/config-converter/src/cli.js" env yaml "${instance_env}" -o "${zowe_yaml}"

    ensure_file_encoding "${zowe_yaml}" "zowe:" "IBM-1047"

    chmod 640 "${zowe_yaml}"
  fi
}

###############################
# Check encoding of a file and convert to IBM-1047 if needed.
#
# Note: usually this is required if the file is supposed to be shell script,
#       which requires to be IBM-1047 encoding.
#
# @param string    file to check and convert
zos_convert_env_dir_file_encoding() {
  file="${1}"

  encoding=$(get_file_encoding "$file")
  echo ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>BEFORE ${file} encoding is ${encoding}"
  cat "$file"
  echo "<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<"
  if [ "${encoding}" != "UNTAGGED" -a "${encoding}" != "IBM-1047" ]; then
    tmpfile="${ZWE_PRIVATE_WORKSPACE_ENV_DIR}/t"
    rm -f "${tmpfile}"
    iconv -f "${encoding}" -t "IBM-1047" "${file}" > "${tmpfile}"
    mv "${tmpfile}" "${file}"
    chmod 640 "${file}"
    echo ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>AFTER ${file}"
    ls -laT "${file}"
    cat "$file"
    echo "<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<"
  fi
}

###############################
# Prepare configuration for current HA instance, and generate backward
# compatible instance.env files from zowe.yaml.
#
# @param string   HA instance ID
generate_instance_env_from_yaml_config() {
  ha_instance="${1}"

  if [ -z "${ZWE_PRIVATE_WORKSPACE_ENV_DIR}" ]; then
    ZWE_PRIVATE_WORKSPACE_ENV_DIR="${ZWE_zowe_workspaceDirectory}/.env"
  fi

  # delete old files to avoid potential issues
  print_formatted_trace "ZWELS" "bin/libs/config.sh,generate_instance_env_from_yaml_config:${LINENO}" "deleting old files under ${ZWE_PRIVATE_WORKSPACE_ENV_DIR}"
  find "${ZWE_PRIVATE_WORKSPACE_ENV_DIR}" -type f -name ".*-${ha_instance}.env" | xargs rm -f
  find "${ZWE_PRIVATE_WORKSPACE_ENV_DIR}" -type f -name ".*-${ha_instance}.json" | xargs rm -f
  find "${ZWE_PRIVATE_WORKSPACE_ENV_DIR}" -type f -name ".zowe.json" | xargs rm -f

  # prepare .zowe.json and .zowe-<ha-id>.json
  print_formatted_trace "ZWELS" "bin/libs/config.sh,generate_instance_env_from_yaml_config:${LINENO}" "config-converter yaml convert --ha ${ha_instance} ${ZWE_CLI_PARAMETER_CONFIG}"
  result=$(node "${ZWE_zowe_runtimeDirectory}/bin/utils/config-converter/src/cli.js" yaml convert --wd "${ZWE_PRIVATE_WORKSPACE_ENV_DIR}" --ha "${ha_instance}" "${ZWE_CLI_PARAMETER_CONFIG}" --verbose)
  code=$?
  print_formatted_trace "ZWELS" "bin/libs/config.sh,generate_instance_env_from_yaml_config:${LINENO}" "- Exit code: ${code}: ${result}"
  if [ ! -f "${ZWE_PRIVATE_WORKSPACE_ENV_DIR}/.zowe.json" ]; then
    print_formatted_error "ZWELS" "bin/libs/config.sh,generate_instance_env_from_yaml_config:${LINENO}" "ZWEL0140E: Failed to translate Zowe configuration (${ZWE_CLI_PARAMETER_CONFIG})."
    exit 140
  fi

  # convert YAML configurations to backward compatible .instance-<ha-id>.env files
  print_formatted_trace "ZWELS" "bin/libs/config.sh,generate_instance_env_from_yaml_config:${LINENO}" "config-converter yaml env --ha ${ha_instance}"
  result=$(node "${ZWE_zowe_runtimeDirectory}/bin/utils/config-converter/src/cli.js" yaml env --wd "${ZWE_PRIVATE_WORKSPACE_ENV_DIR}" --ha "${ha_instance}" --verbose)
  code=$?
  print_formatted_trace "ZWELS" "bin/libs/config.sh,generate_instance_env_from_yaml_config:${LINENO}" "- Exit code: ${code}: ${result}"
  if [ ! -f "${ZWE_PRIVATE_WORKSPACE_ENV_DIR}/.instance-${ha_instance}.env" ]; then
    print_formatted_error "ZWELS" "bin/libs/config.sh,generate_instance_env_from_yaml_config:${LINENO}" "ZWEL0140E: Failed to translate Zowe configuration (${ZWE_CLI_PARAMETER_CONFIG})."
    exit 140
  fi
}

# check and sanitize ZWE_CLI_PARAMETER_HA_INSTANCE
sanitize_ha_instance_id() {
  # ignore default value passed from ZWESLSTC
  if [ "${ZWE_CLI_PARAMETER_HA_INSTANCE}" = "{{ha_instance_id}}" -o "${ZWE_CLI_PARAMETER_HA_INSTANCE}" = "__ha_instance_id__" ]; then
    ZWE_CLI_PARAMETER_HA_INSTANCE=
  fi
  if [ -z "${ZWE_CLI_PARAMETER_HA_INSTANCE}" ]; then
    ZWE_CLI_PARAMETER_HA_INSTANCE=$(get_sysname)
  fi
  # sanitize instance id
  ZWE_CLI_PARAMETER_HA_INSTANCE=$(echo "${ZWE_CLI_PARAMETER_HA_INSTANCE}" | lower_case | sanitize_alphanum)
}

###############################
# Load environment variables used by components
#
# NOTE: all environment variables used/defined by Zowe should be ensured in this function.
#       "zwe internal start prepare" is the only special case where we may need to define some variables before calling
#       this function. The reason is to properly prepare the directories, logging, etc.
load_environment_variables() {
  component_id="${1}"

  # check and sanitize ZWE_CLI_PARAMETER_HA_INSTANCE
  sanitize_ha_instance_id

  if [ -z "${ZWE_zowe_workspaceDirectory}" ]; then
    ZWE_zowe_workspaceDirectory=$(read_yaml "${ZWE_CLI_PARAMETER_CONFIG}" '.zowe.workspaceDirectory')
    if [ -z "${ZWE_zowe_workspaceDirectory}" ]; then
      print_error_and_exit "Error ZWEL0157E: Zowe workspace directory (zowe.workspaceDirectory) is not defined in Zowe YAML configuration file." "" 157
    fi
  fi

  if [ -z "${ZWE_VERSION}" ]; then
    export ZWE_VERSION=$(shell_read_json_config "${ZWE_zowe_runtimeDirectory}/manifest.json" 'version' 'version')
  fi

  # we must have $ZWE_zowe_workspaceDirectory at this point
  if [ -f "${ZWE_zowe_workspaceDirectory}/.init-for-container" ]; then
    export ZWE_RUN_IN_CONTAINER=true
  fi

  # these are already set in prepare stage, re-ensure for start
  export ZWE_PRIVATE_WORKSPACE_ENV_DIR="${ZWE_zowe_workspaceDirectory}/.env"
  export ZWE_STATIC_DEFINITIONS_DIR="${ZWE_zowe_workspaceDirectory}/api-mediation/api-defs"
  export ZWE_GATEWAY_SHARED_LIBS="${ZWE_zowe_workspaceDirectory}/gateway/sharedLibs/"
  export ZWE_DISCOVERY_SHARED_LIBS="${ZWE_zowe_workspaceDirectory}/discovery/sharedLibs/"

  # now we can load all variables
  if [ -n "${component_id}" -a -f "${ZWE_PRIVATE_WORKSPACE_ENV_DIR}/${component_id}/.instance-${ZWE_CLI_PARAMETER_HA_INSTANCE}.env" ]; then
    source_env "${ZWE_PRIVATE_WORKSPACE_ENV_DIR}/${component_id}/.instance-${ZWE_CLI_PARAMETER_HA_INSTANCE}.env"
  elif [ -f "${ZWE_PRIVATE_WORKSPACE_ENV_DIR}/.instance-${ZWE_CLI_PARAMETER_HA_INSTANCE}.env" ]; then
    source_env "${ZWE_PRIVATE_WORKSPACE_ENV_DIR}/.instance-${ZWE_CLI_PARAMETER_HA_INSTANCE}.env"
  else
    print_error_and_exit "Error ZWEL0112E: Zowe runtime environment must be prepared first with \"zwe internal start prepare\" command." "" 112
  fi

  # ZWE_DISCOVERY_SERVICES_LIST should have been prepared in zowe-install-packaging-tools and had been sourced.

  # overwrite ZWE_PRIVATE_LOG_LEVEL_ZWELS with zowe.launchScript.logLevel config in YAML
  export ZWE_PRIVATE_LOG_LEVEL_ZWELS="$(echo "${ZWE_zowe_launchScript_logLevel}" | upper_case)"

  # generate other variables
  export ZWE_INSTALLED_COMPONENTS="$(find_all_installed_components)"
  export ZWE_ENABLED_COMPONENTS="$(find_all_enabled_components)"
  export ZWE_LAUNCH_COMPONENTS="$(find_all_launch_components)"

  # ZWE_DISCOVERY_SERVICES_LIST should have been prepared in zowe-install-packaging-tools

  if [ "${ZWE_RUN_IN_CONTAINER}" = "true" ]; then
    prepare_container_runtime_environments
  fi
}
