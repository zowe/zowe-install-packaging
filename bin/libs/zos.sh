#!/bin/sh

#######################################################################
# This program and the accompanying materials are made available
# under the terms of the Eclipse Public License v2.0 which
# accompanies this distribution, and is available at
# https://www.eclipse.org/legal/epl-v20.html
#
# SPDX-License-Identifier: EPL-2.0
#
# Copyright Contributors to the Zowe Project.
#######################################################################

tso_command() {
  message="- tsocmd $@"
  print_debug "${message}"
  result=$(tsocmd "$@" 2>&1)
  code=$?
  if [ ${code} -eq 0 ]; then
    print_debug "  * Succeeded"
    print_trace "  * Exit code: ${code}"
    print_trace "  * Output:"
    if [ -n "${result}" ]; then
      print_trace "$(padding_left "${result}" "    ")"
    fi
  else
    print_debug "  * Failed"
    print_error "  * Exit code: ${code}"
    print_error "  * Output:"
    if [ -n "${result}" ]; then
      print_error "$(padding_left "${result}" "    ")"
    fi
  fi

  echo "${result}"

  return ${code}
}

operator_command() {
  cmd="${1}"

  opercmd=${ZWE_zowe_runtimeDirectory}/bin/utils/opercmd.rex

  message="- opercmd ${cmd}"
  print_debug "${message}"
  result=$("${opercmd}" "${cmd}" 2>&1)
  code=$?
  if [ ${code} -eq 0 ]; then
    print_debug "  * Succeeded"
    print_trace "  * Exit code: ${code}"
    print_trace "  * Output:"
    if [ -n "${result}" ]; then
      print_trace "$(padding_left "${result}" "    ")"
    fi
  else
    print_debug "  * Failed"
    print_error "  * Exit code: ${code}"
    print_error "  * Output:"
    if [ -n "${result}" ]; then
      print_error "$(padding_left "${result}" "    ")"
    fi
  fi

  echo "${result}"

  return ${code}
}

verify_generated_jcl() {
  jcllib=$(read_yaml "${ZWE_CLI_PARAMETER_CONFIG}" ".zowe.setup.dataset.jcllib")
  # read JCL library and validate
  does_jcl_exist=$(is_data_set_exists "${jcllib}")
  if [ -z "${does_jcl_exist}" ]; then
    result=$(zwecli_inline_execute_command init generate)
  fi

  # should be created, but may take time to discover.
  if [ -z "${does_jcl_exist}" ]; then
    does_jcl_exist=
    for secs in 1 5 10 ; do
      does_jcl_exist=$(is_data_set_exists "${jcllib}")
      if [ -z "${does_jcl_exist}" ]; then
        sleep ${secs}
      else
        break
      fi
    done

    if [ -z "${does_jcl_exist}" ]; then
      return 1
    fi
  fi
  echo "${jcllib}"
  return 0
}
