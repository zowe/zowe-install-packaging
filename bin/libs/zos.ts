/*
  This program and the accompanying materials are made available
  under the terms of the Eclipse Public License v2.0 which
  accompanies this distribution, and is available at
  https://www.eclipse.org/legal/epl-v20.html
 
  SPDX-License-Identifier: EPL-2.0
 
  Copyright Contributors to the Zowe Project.
*/

import * as std from 'cm_std';
import * as os from 'cm_os';
import * as common from './common';
import * as shell from './shell';
import * as stringlib from './string';
import * as zosDataset from './zos-dataset';
import * as initGenerate from '../commands/init/generate/index';

export function tsoCommand(...args:string[]): { rc: number, out: string } {
  let message = "tsocmd " + '"' + args.join(' ') + '"';
  common.printDebug('- '+message);
  //we echo at the end to avoid a configmgr quirk where trying to read stdout when empty can hang waiting for bytes
  const result = shell.execOutSync('sh', '-c', `${message} 2>&1 && echo '.'`);
  if (result.rc == 0) {
    common.printDebug("  * Succeeded");
    common.printTrace(`  * Exit code: ${result.rc}`);
    common.printTrace("  * Output:");
    if (result.out) {
      result.out = result.out.substring(0, result.out.length - 1);
      common.printTrace(stringlib.paddingLeft(result.out, "    "));
    }
  } else {
    common.printDebug("  * Failed");
    common.printError(`  * Exit code: ${result.rc}`);
    common.printError("  * Output:");
    if (result.out) {
      common.printError(stringlib.paddingLeft(result.out, "    "));
    }
  }
  return { rc: result.rc, out: result.out ? result.out : '' };
}

export function operatorCommand(command: string): { rc: number, out: string } {
  const opercmd=std.getenv('ZWE_zowe_runtimeDirectory')+'/bin/utils/opercmd.rex';

  let message=`- opercmd ${command}`;
  common.printDebug(message);
  //we echo at the end to avoid a configmgr quirk where trying to read stdout when empty can hang waiting for bytes
  const result = shell.execOutSync('sh', '-c', `${opercmd} "${command}" 2>&1 && echo '.'`);
  if (result.rc == 0) {
    common.printDebug("  * Succeeded");
    common.printTrace(`  * Exit code: ${result.rc}`);
    common.printTrace("  * Output:");
    if (result.out) {
      common.printTrace(stringlib.paddingLeft(result.out, "    "));
    }
  } else {
    common.printDebug("  * Failed");
    common.printError(`  * Exit code: ${result.rc}`);
    common.printError("  * Output:");
    if (result.out) {
      common.printError(stringlib.paddingLeft(result.out, "    "));
    }
  }
  //we strip the '.' we added above
  return { rc: result.rc, out: result.out ? result.out.substring(0, result.out.length-1) : '' };
}

export function verifyGeneratedJcl(config:any): string {
  const jcllib = config.zowe.setup.dataset.jcllib;
  if (!jcllib) {
    return undefined;
  }
  const expectedMember = jcllib+'(ZWEIMVS)';
  // read JCL library and validate using expected member ZWEIMVS (init mvs command)
  let doesJclExist: boolean = zosDataset.isDatasetExists(expectedMember);
  if (!doesJclExist) {
    initGenerate.execute();
  }

  // should be created, but may take time to discover.
  if (!doesJclExist) {
    const interval = [1,5,10,30];
    for (let i = 0; i < interval.length; i++) {
      let secs = interval[i];
      doesJclExist=zosDataset.isDatasetExists(expectedMember);
      if (!doesJclExist) {
        os.sleep(secs*1000);
      } else {
        break;
      }
    }

    if (!doesJclExist) {      
      return undefined;
    }
  }
  return jcllib;
}
