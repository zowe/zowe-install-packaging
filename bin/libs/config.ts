/*
  This program and the accompanying materials are made available
  under the terms of the Eclipse Public License v2.0 which
  accompanies this distribution, and is available at
  https://www.eclipse.org/legal/epl-v20.html
 
  SPDX-License-Identifier: EPL-2.0
 
  Copyright Contributors to the Zowe Project.
*/

import * as std from 'std';
import * as os from 'os';
import * as zos from 'zos';
import * as zosfs from './zosfs';
import * as fs from './fs';
import * as common from './common';
import * as stringlib from './string';
import * as shell from './shell';
import * as sys from './sys';
import * as component from './component';
import * as container from './container';
import * as varlib from './var';

const runtimeDirectory=std.getenv('ZWE_zowe_runtimeDirectory');
const extensionDirectory=std.getenv('ZWE_zowe_extensionDirectory');
const workspaceDirectory=std.getenv('ZWE_zowe_workspaceDirectory');
let parameterConfig = std.getenv('ZWE_CLI_PARAMETER_CONFIG');
const configPath = (parameterConfig && !parameterConfig.startsWith('FILE(')) ? `FILE(${parameterConfig})` : parameterConfig;
let configLoaded = false;

const COMMON_SCHEMA = `${runtimeDirectory}/schemas/server-common.json`;
const ZOWE_SCHEMA = `${runtimeDirectory}/schemas/zowe-yaml-schema.json`;
const ZOWE_SCHEMA_ID = 'https://zowe.org/schemas/v2/server-base';
const ZOWE_SCHEMA_SET=`${ZOWE_SCHEMA}:${COMMON_SCHEMA}`;

function configMgrFailMessage(name:string) {
  common.printError('Failed to init config '+name);
  std.exit(1);
}


export function getZoweConfig(): any {
  if (configLoaded) {
    return common.CONFIG_MGR.getConfigData('zowe-server-base');
  }
  
  if (configPath) {
    let status;

    if ((status = common.CONFIG_MGR.addConfig('zowe-server-base'))) {
      common.printError(`Could not add config for ${configPath}, status=${status}`);
      std.exit(1);
    }

    if ((status = common.CONFIG_MGR.loadSchemas('zowe-server-base', ZOWE_SCHEMA_SET))) {
      common.printError(`Could not load schemas ${ZOWE_SCHEMA_SET} for configs ${configPath}, status=${status}`);
      std.exit(1);
    }

    if ((status = common.CONFIG_MGR.setConfigPath('zowe-server-base', configPath))) {
      common.printError(`Could not set config path for ${configPath}, status=${status}`);
      std.exit(1);
    }

    if ((status = common.CONFIG_MGR.loadConfiguration('zowe-server-base'))) {
      common.printError(`Could not load config for ${configPath}, status=${status}`);
      std.exit(1);
    }

    let validation = common.CONFIG_MGR.validate('zowe-server-base');
    if (validation.ok){
      if (validation.exceptions){
        common.printError(`Validation of ${configPath} against schema ${ZOWE_SCHEMA_ID} found invalid JSON Schema data`);
        for (let i=0; i<validation.exceptions.length; i++){
          common.printError("    "+validation.exceptions[i]);
        }
        std.exit(1);
      } else {
        configLoaded = true;
        return common.CONFIG_MGR.getConfigData('zowe-server-base');
      }
    } else {
      common.printError(`Error occurred on validation of ${configPath} against schema ${ZOWE_SCHEMA_ID} `);
      std.exit(1);
    }
  } else {
    common.printError(`Server config path not given`);
    std.exit(1);
  }  
}

// Convert instance.env to zowe.yaml file
export function convertInstanceEnvToYaml(instanceEnv: string, zoweYaml?: string) {
  // we need node for following commands
  node.ensureNodeIsOnPath();

  if (!zoweYaml) {
    shell.execSync('node', `${std.getenv('ROOT_DIR')}/bin/utils/config-converter/src/cli.js" env yaml "${instanceEnv}`);
  } else {
    shell.execSync('node', `${std.getenv('ROOT_DIR')}/bin/utils/config-converter/src/cli.js" env yaml "${instanceEnv}" -o "${zoweYaml}`);

    zosfs.ensureFileEncoding(zoweYaml, "zowe:", "IBM-1047");

    shell.execSync('chmod', `640 "${zoweYaml}"`);
  }
}

//////////////////////////////////////////////////////////////
// Check encoding of a file and convert to IBM-1047 if needed.
//
// Note: usually this is required if the file is supposed to be shell script,
//       which requires to be IBM-1047 encoding.
//
export function zosConvertEnvDirFileEncoding(file: string) {

  const encoding=zosfs.getFileEncoding(file);
  console.log( ">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>BEFORE ${file} encoding is ${encoding}");
  console.log(std.loadFile(file));
  console.log( "<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
  if (encoding && encoding != "UNTAGGED" && encoding != "IBM-1047") {
    const tmpfile=`${std.getenv('ZWE_PRIVATE_WORKSPACE_ENV_DIR')}/t`;
    os.remove(tmpfile);
    shell.execSync(`iconv`, `-f "${encoding}" -t "IBM-1047" "${file}" > "${tmpfile}"`);
    os.rename(tmpfile, file);
    shell.execSync('chmod', `640 "${file}"`);
    console.log(`>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>AFTER ${file}`);
    const lsReturn = shell.execOutSync('ls', `-laT "${file}"`);
    console.log(lsReturn.out);
    console.log(std.loadFile(file));
    console.log("<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<");
  }
}

// Prepare configuration for current HA instance, and generate backward
// compatible instance.env files from zowe.yaml.
//
export function generateInstanceEnvFromYamlConfig(haInstance: string) {
  const workspaceDirectory=std.getenv('ZWE_zowe_workspaceDirectory');
  const runtimeDirectory=std.getenv('ZWE_zowe_runtimeDirectory');
  const zweCliParameterConfig = std.getenv('ZWE_CLI_PARAMETER_CONFIG');
  let zwePrivateWorkspaceEnvDir = std.getenv('zwePrivateWorkspaceEnvDir');
  if (!zwePrivateWorkspaceEnvDir) {
    zwePrivateWorkspaceEnvDir=`${workspaceDirectory}/.env`
    std.setenv('zwePrivateWorkspaceEnvDir', zwePrivateWorkspaceEnvDir);
  }

  // delete old files to avoid potential issues
  common.printFormattedTrace( "ZWELS", "bin/libs/config.sh,generate_instance_env_from_yaml_config", "deleting old files under ${zwePrivateWorkspaceEnvDir}");
  shell.execSync('sh', `find "${zwePrivateWorkspaceEnvDir}" -type f -name ".*-${haInstance}.env" | xargs rm -f`);
  shell.execSync('sh', `find "${zwePrivateWorkspaceEnvDir}" -type f -name ".*-${haInstance}.json" | xargs rm -f`);
  shell.execSync('sh', `find "${zwePrivateWorkspaceEnvDir}" -type f -name ".zowe.json" | xargs rm -f`);

  // prepare .zowe.json and .zowe-<ha-id>.json
  common.printFormattedTrace("ZWELS", "bin/libs/config.sh,generate_instance_env_from_yaml_config", `config-converter yaml convert --ha ${haInstance} ${zweCliParameterConfig}`);
  let result = shell.execOutSync('node', `${runtimeDirectory}/bin/utils/config-converter/src/cli.js" yaml convert --wd "${zwePrivateWorkspaceEnvDir}" --ha "${haInstance}" "${zweCliParameterConfig}" --verbose"`);

  common.printFormattedTrace("ZWELS", "bin/libs/config.sh,generate_instance_env_from_yaml_config", `- Exit code: ${result.rc}: ${result.out}`);
  if ( !fs.fileExists(`${zwePrivateWorkspaceEnvDir}/.zowe.json`)) {
    common.printFormattedError( "ZWELS", "bin/libs/config.sh,generate_instance_env_from_yaml_config", `ZWEL0140E: Failed to translate Zowe configuration (${zweCliParameterConfig}).`);
    std.exit(140);
  }

  // convert YAML configurations to backward compatible .instance-<ha-id>.env files
  common.printFormattedTrace("ZWELS", "bin/libs/config.sh,generate_instance_env_from_yaml_config", `config-converter yaml env --ha ${haInstance}`);
  result=shell.execOutSync('node', `"${runtimeDirectory}/bin/utils/config-converter/src/cli.js" yaml env --wd "${zwePrivateWorkspaceEnvDir}" --ha "${haInstance}" --verbose`);

  common.printFormattedTrace("ZWELS", "bin/libs/config.sh,generate_instance_env_from_yaml_config", `- Exit code: ${result.rc}: ${result.out}`);
  
  if (!fs.fileExists("${zwePrivateWorkspaceEnvDir}/.instance-${haInstance}.env")) {
    common.printFormattedError("ZWELS", "bin/libs/config.sh,generate_instance_env_from_yaml_config", `ZWEL0140E: Failed to translate Zowe configuration (${zweCliParameterConfig}).`);
    std.exit(140);
  }
}


// check and sanitize ZWE_CLI_PARAMETER_HA_INSTANCE
export function sanitizeHaInstanceId() {
  // ignore default value passed from ZWESLSTC
  let zweCliParameterHaInstance = std.getenv('ZWE_CLI_PARAMETER_HA_INSTANCE');
  if (zweCliParameterHaInstance == "{{ha_instance_id}}" || zweCliParameterHaInstance == "__ha_instance_id__") {
    std.unsetenv('ZWE_CLI_PARAMETER_HA_INSTANCE');
    zweCliParameterHaInstance=undefined;
  }
  if (!zweCliParameterHaInstance) {
    zweCliParameterHaInstance=sys.getSysname();
  }
  // sanitize instance id
  zweCliParameterHaInstance=stringlib.sanitizeAlphanum(zweCliParameterHaInstance.toLowerCase());
  std.setenv('ZWE_CLI_PARAMETER_HA_INSTANCE', zweCliParameterHaInstance);
}

export function applyEnviron(environ: any): void {
  let keys = Object.keys(environ);
  keys.forEach(function(key:string) {
    std.setenv(key, environ[key]);
  });
}

//////////////////////////////////////////////////////////////
// Load environment variables used by components
//
// NOTE: all environment variables used/defined by Zowe should be ensured in this function.
//       "zwe internal start prepare" is the only special case where we may need to define some variables before calling
//       this function. The reason is to properly prepare the directories, logging, etc.
export function loadEnvironmentVariables(componentId: string) {

  // check and sanitize zweCliParameterHaInstance
  sanitizeHaInstanceId();
  let workspaceDirectory=std.getenv('ZWE_zowe_workspaceDirectory');
  if (!workspaceDirectory) {
    let zoweConfig = getZoweConfig();
    workspaceDirectory=zoweConfig.zowe.workspaceDirectory;
    std.setenv('ZWE_zowe_workspaceDirectory',workspaceDirectory);
  }

  if (!std.getenv('ZWE_VERSION')) {
    const runtimeManifestString=std.loadFile(`${runtimeDirectory}/manifest.json`);
    if (runtimeManifestString) {
      std.setenv('ZWE_VERSION', JSON.parse(runtimeManifestString).version);
    }
  }

  // we must have $workspaceDirectory at this point
  if (fs.fileExists(`${workspaceDirectory}/.init-for-container`)) {
    std.setenv('ZWE_RUN_IN_CONTAINER','true');
  }

  // these are already set in prepare stage, re-ensure for start
  std.setenv('ZWE_PRIVATE_WORKSPACE_ENV_DIR', `${workspaceDirectory}/.env`);
  std.setenv('ZWE_STATIC_DEFINITIONS_DIR', `${workspaceDirectory}/api-mediation/api-defs`);
  std.setenv('ZWE_GATEWAY_SHARED_LIBS', `${workspaceDirectory}/gateway/sharedLibs/`);
  std.setenv('ZWE_DISCOVERY_SHARED_LIBS', `${workspaceDirectory}/discovery/sharedLibs/`);

  // now we can load all variables
  let zweCliParameterHaInstance=std.getenv('ZWE_CLI_PARAMETER_HA_INSTANCE');
  let zwePrivateWorkspaceEnvDir=std.getenv('ZWE_PRIVATE_WORKSPACE_ENV_DIR');    
  if (componentId && fs.fileExists(`${workspaceDirectory}/.env/${componentId}/.instance-${zweCliParameterHaInstance}.env`)) {
    varlib.sourceEnv(`${zwePrivateWorkspaceEnvDir}/${componentId}/.instance-${zweCliParameterHaInstance}.env`);
  } else if (fs.fileExists(`${zwePrivateWorkspaceEnvDir}/.instance-${zweCliParameterHaInstance}.env`)) {
    varlib.sourceEnv(`${zwePrivateWorkspaceEnvDir}/.instance-${zweCliParameterHaInstance}.env`);
  } else {
    common.printErrorAndExit( "Error ZWEL0112E: Zowe runtime environment must be prepared first with \"zwe internal start prepare\" command.", undefined, 112);
  }

  // ZWE_DISCOVERY_SERVICES_LIST should have been prepared in zowe-install-packaging-tools and had been sourced.

  // overwrite ZWE_PRIVATE_LOG_LEVEL_ZWELS with zowe.launchScript.logLevel config in YAML
  let logLevel =  std.getenv('ZWE_zowe_launchScript_logLevel');
  if (logLevel) {
    std.setenv('ZWE_PRIVATE_LOG_LEVEL_ZWELS', logLevel.toUpperCase());
  }
  // generate other variables
  std.setenv('ZWE_INSTALLED_COMPONENTS', component.findAllInstalledComponents());
  std.setenv('ZWE_ENABLED_COMPONENTS', component.findAllEnabledComponents());
  std.setenv('ZWE_LAUNCH_COMPONENTS', component.findAllLaunchComponents());

  // ZWE_DISCOVERY_SERVICES_LIST should have been prepared in zowe-install-packaging-tools

  if (std.getenv('ZWE_RUN_IN_CONTAINER') == "true") {
    container.prepareContainerRuntimeEnvironments();
  }
  return std.getenviron();
}
