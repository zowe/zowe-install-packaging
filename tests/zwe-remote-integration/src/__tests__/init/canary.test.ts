/*
 * This program and the accompanying materials are made available under the terms of the
 * Eclipse Public License v2.0 which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Copyright Contributors to the Zowe Project.
 */

import { RemoteTestRunner } from '../../zos/RemoteTestRunner';
import { ZoweConfig } from '../../config/ZoweConfig';

/**
 * Make sure the test suite and remote system are working as expected
 */
const testSuiteName = 'canary';
describe(testSuiteName, () => {
  let testRunner: RemoteTestRunner;
  beforeAll(() => {
    testRunner = new RemoteTestRunner('canary');
  });

  afterAll(() => {
    testRunner.shutdown();
  });

  it('run echo', async () => {
    const cfgYaml = ZoweConfig.getZoweYaml();
    cfgYaml.java.home = '/ZOWE/node/J21.0_64/';
    const result = await testRunner.runRaw('echo "hi"');
    expect(result.rc).toBe(0);
    expect(result.stdout).not.toBeNull();
    expect(result.stdout).toMatchSnapshot();
  });

  it('run zwe help', async () => {
    const cfgYaml = ZoweConfig.getZoweYaml();
    cfgYaml.java.home = '/ZOWE/node/J21.0_64/';
    const result = await testRunner.runZweTest(cfgYaml, '--help');
    expect(result.rc).toBe(100); // 100 is expected...
    expect(result.stdout).not.toBeNull();
    expect(result.stdout).toMatchSnapshot();
  });
});
