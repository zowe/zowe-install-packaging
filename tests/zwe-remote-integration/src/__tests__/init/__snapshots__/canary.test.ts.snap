// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`canary run echo 1`] = `"hi"`;

exports[`canary run zwe help 1`] = `
"zwe 

------------------
Synopsis
    zwe  [sub-command [sub-command]...] [parameter [parameter]...]

------------------
Description
    A command line utility helps you managing Zowe instance.
    
    You can issue --help or -h to find information for all commands it supports.

------------------
Global parameters
    --help|-h  (optional)
        Display this help. 

    --debug|--verbose|-v  (optional)
        Enable verbose mode. 

    --trace|-vv  (optional)
        Enable trace level debug mode. 

    --silent|-s  (optional)
        Do not display messages to standard output. 

    --log-dir|--log|-l string (optional)
        Write logs to this directory. 

    --config|-c string (optional)
        Path to Zowe configuration zowe.yaml file. 

    --configmgr  (optional)
        Enable use of configmgr capabilities. 

------------------
Available sub-command(s)
    - certificate
    - components
    - config
    - diagnose
    - init
    - install
    - internal
    - migrate
    - sample
    - start
    - stop
    - support
    - version

------------------
Example(s)
    zwe install -h
    
    zwe init --allow-overwrite --config /path/to/zowe.yaml
    
    zwe start -c /path/to/zowe.yaml"
`;
